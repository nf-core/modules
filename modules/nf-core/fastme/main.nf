process FASTME {
    tag "$infile"
    label 'process_medium'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/fastme:2.1.6.1--hec16e2b_1':
        'biocontainers/fastme:2.1.6.1--hec16e2b_1' }"

    input:
    tuple val(meta), path(infile), path(initial_tree)

    output:
    tuple val(meta), path("*.nwk")       , emit: nwk
    tuple val(meta), path("*_stat.txt")  , emit: stats
    tuple val(meta), path("*.matrix.phy"), emit: matrix    , optional: true
    tuple val(meta), path("*.bootstrap") , emit: bootstrap , optional: true
    path "versions.yml" , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args    = task.ext.args ?: ''
    def prefix  = task.ext.prefix ?: infile
    def initarg = initial_tree ? "-u $initial_tree" : ''
    def matarg  = task.ext.args =~ "-O" ? "-O ${prefix}.matrix.phy" : ''
    def bootarg = task.ext.args =~ "-B" ? "-B ${prefix}.bootstrap" : ''
    """
    fastme \\
        $args \\
        -i $infile \\
        $initarg \\
        -o ${prefix}.nwk \\
        $matarg \\
        $bootarg \\
        -T $task.cpus


    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        fastme: \$(fastme --version |& sed '1!d ; s/FastME //')
    END_VERSIONS
    """

    stub:
    def args   = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: infile
    def mat    = task.ext.args =~ "-O" ? "touch ${prefix}.matrix.phy" : ''
    def boot   = task.ext.args =~ "-B" ? "touch ${prefix}.bootstrap" : ''
    """
    touch ${prefix}.nwk
    touch ${prefix}_stat.txt
    $mat
    $boot

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        fastme: \$(fastme --version |& sed '1!d ; s/FastME //')
    END_VERSIONS
    """
}
