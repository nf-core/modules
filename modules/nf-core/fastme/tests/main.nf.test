nextflow_process {

    name "Test Process FASTME"
    script "../main.nf"
    process "FASTME"

    tag "modules"
    tag "modules_nfcore"
    tag "fastme"
    tag "tcoffee/seqreformat"
    tag "famsa/guidetree"

    test("setoxin - phylip - basic") {

        config "./main.config"

        setup {
            run("TCOFFEE_SEQREFORMAT") {
                script "../../tcoffee/seqreformat/main.nf"
                process {
                    """
                    input[0] = [ [ id: "test" ],
                        file("https://raw.githubusercontent.com/nf-core/test-datasets/multiplesequencealign/testdata/setoxin.ref", checkIfExists: true)
                    ]
                    """
                }
            }
        }

        when {
            process {
                """
                input[0] = TCOFFEE_SEQREFORMAT.out.formatted_file
                    .map { meta, aln -> [meta, aln, []] }
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("setoxin - phylip - with_tree") {

        config "./main.config"

        setup {
            run("TCOFFEE_SEQREFORMAT") {
                script "../../tcoffee/seqreformat/main.nf"
                process {
                    """
                    input[0] = [ [ id: "test" ],
                        file("https://raw.githubusercontent.com/nf-core/test-datasets/multiplesequencealign/testdata/setoxin.ref", checkIfExists: true)
                    ]
                    """
                }
            }
            run("FAMSA_GUIDETREE") {
                script "../../famsa/guidetree/main.nf"
                process {
                    """
                    input[0] = [ [ id: "test" ],
                                file("https://raw.githubusercontent.com/nf-core/test-datasets/multiplesequencealign/testdata/setoxin.ref", checkIfExists: true)
                            ]

                    """
                }
            }
        }

        when {
            process {
                """
                input[0] = TCOFFEE_SEQREFORMAT.out.formatted_file
                    .join(FAMSA_GUIDETREE.out.tree, by: 0)
                    .map { meta, aln, tree -> [meta, aln, tree] }
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }

    }

    test("setoxin - phylip - bootstrap") {

        config "./optionals.config"

        setup {
            run("TCOFFEE_SEQREFORMAT") {
                script "../../tcoffee/seqreformat/main.nf"
                process {
                    """
                    input[0] = [ [ id: "test" ],
                        file("https://raw.githubusercontent.com/nf-core/test-datasets/multiplesequencealign/testdata/setoxin.ref", checkIfExists: true)
                    ]
                    """
                }
            }
        }

        when {
            process {
                """
                input[0] = TCOFFEE_SEQREFORMAT.out.formatted_file
                    .map { meta, aln -> [meta, aln, []] }
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert path(process.out.nwk[0][1]).text.contains("1atx:") },
                { assert path(process.out.matrix[0][1]).text.contains("1apf") },
                { assert path(process.out.bootstrap[0][1]).text.contains("1atx:") },
                { assert snapshot(path(process.out.stats[0][1]).readLines()[0..12]).match("stats_boot") },
                { assert snapshot(process.out.versions).match("versions") }
            )
        }
    }

    test("setoxin - phylip - stub") {

        config "./main.config"
        options "-stub"

        when {
            process {
                """
                input[0] = [ [ id: "test" ],
                    file("https://raw.githubusercontent.com/nf-core/test-datasets/multiplesequencealign/testdata/setoxin.ref", checkIfExists: true),
                    []
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }
}