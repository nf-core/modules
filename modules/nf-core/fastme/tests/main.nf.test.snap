{
    "setoxin - phylip - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "setoxin.ref.nwk:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "setoxin.ref_stat.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    
                ],
                "3": [
                    
                ],
                "4": [
                    "versions.yml:md5,0e7f28ae349efffa1ef75c2279e975b6"
                ],
                "bootstrap": [
                    
                ],
                "matrix": [
                    
                ],
                "nwk": [
                    [
                        {
                            "id": "test"
                        },
                        "setoxin.ref.nwk:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "stats": [
                    [
                        {
                            "id": "test"
                        },
                        "setoxin.ref_stat.txt:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,0e7f28ae349efffa1ef75c2279e975b6"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.0"
        },
        "timestamp": "2024-03-19T10:03:04.842045"
    },
    "versions": {
        "content": [
            [
                "versions.yml:md5,0e7f28ae349efffa1ef75c2279e975b6"
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.0"
        },
        "timestamp": "2024-03-19T10:02:58.72899"
    },
    "stats_boot": {
        "content": [
            [
                "",
                " - FastME 2.1.6.1 - ",
                "",
                "",
                "Papers to be cited:",
                "",
                "FastME 2.0 - A comprehensive, accurate and fast distance-based phylogeny inference program.",
                "\tVincent Lefort, Richard Desper and Olivier Gascuel,",
                "\tMolecular Biology and Evolution 32(10), 2798-800, 2015.",
                "BIONJ algorithm:",
                "\tGascuel O. 1997. BIONJ: an improved version of the NJ algorithm based on a simple model of sequence data.",
                "\tMolecular Biology and Evolution, 14(7):685-695",
                ""
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.0"
        },
        "timestamp": "2024-03-19T10:09:35.813028"
    },
    "setoxin - phylip - with_tree": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.txt.nwk:md5,cbd6a41704951c56512f2f755dc13d4e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test.txt_fastme_stat.txt:md5,de3629be9e561cd78286bc565036a1d9"
                    ]
                ],
                "2": [
                    
                ],
                "3": [
                    
                ],
                "4": [
                    "versions.yml:md5,0e7f28ae349efffa1ef75c2279e975b6"
                ],
                "bootstrap": [
                    
                ],
                "matrix": [
                    
                ],
                "nwk": [
                    [
                        {
                            "id": "test"
                        },
                        "test.txt.nwk:md5,cbd6a41704951c56512f2f755dc13d4e"
                    ]
                ],
                "stats": [
                    [
                        {
                            "id": "test"
                        },
                        "test.txt_fastme_stat.txt:md5,de3629be9e561cd78286bc565036a1d9"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,0e7f28ae349efffa1ef75c2279e975b6"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.0"
        },
        "timestamp": "2024-03-19T10:02:51.77025"
    },
    "setoxin - phylip - basic": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.txt.nwk:md5,72ef94af973b93bec264149ae4abafb3"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test.txt_fastme_stat.txt:md5,b8cfaff0c62868a8dea2614f09d0e5af"
                    ]
                ],
                "2": [
                    
                ],
                "3": [
                    
                ],
                "4": [
                    "versions.yml:md5,0e7f28ae349efffa1ef75c2279e975b6"
                ],
                "bootstrap": [
                    
                ],
                "matrix": [
                    
                ],
                "nwk": [
                    [
                        {
                            "id": "test"
                        },
                        "test.txt.nwk:md5,72ef94af973b93bec264149ae4abafb3"
                    ]
                ],
                "stats": [
                    [
                        {
                            "id": "test"
                        },
                        "test.txt_fastme_stat.txt:md5,b8cfaff0c62868a8dea2614f09d0e5af"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,0e7f28ae349efffa1ef75c2279e975b6"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.0"
        },
        "timestamp": "2024-03-19T10:02:44.598308"
    }
}